server/service/src/invoice/outbound_return/insert/mod.rs

```diff
@@ -0,0 +1,377 @@
+        insert_stock_out_line, InsertStockOutLine, InsertStockOutLineError, StockOutType,
+    },
+    service_provider::ServiceContext,
+};
+pub mod generate;
+pub mod validate;
+use generate::generate;
+use validate::validate;
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct InsertOutboundReturn {
+    pub id: String,
+    pub other_party_id: String,
+    pub outbound_return_lines: Vec<InsertOutboundReturnLine>,
+}
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct InsertOutboundReturnLine {
+    pub id: String,
+    pub stock_line_id: String,
+    pub number_of_packs: f64,
+    pub reason_id: String,
+    pub note: String,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum InsertOutboundReturnError {
+    InvoiceAlreadyExists,
+    // Name validation
+    OtherPartyNotASupplier,
+    OtherPartyNotVisible,
+    OtherPartyDoesNotExist,
+    // Internal
+    NewlyCreatedInvoiceDoesNotExist,
+    DatabaseError(RepositoryError),
+    // Line Errors
+    LineInsertError {
+        line_id: String,
+        error: InsertStockOutLineError,
+    },
+    LineReturnReasonUpdateError {
+        line_id: String,
+        error: RepositoryError,
+    },
+}
+
+type OutError = InsertOutboundReturnError;
+
+pub fn insert_outbound_return(
+    ctx: &ServiceContext,
+    input: InsertOutboundReturn,
+) -> Result<Invoice, OutError> {
+    let outbound_return: Invoice = ctx
+        .connection
+        .transaction_sync(|connection| {
+            let other_party = validate(connection, &ctx.store_id, &input)?;
+            let new_invoice = generate(
+                connection,
+                &ctx.store_id,
+                &ctx.user_id,
+                input.clone(),
+                other_party,
+            )?;
+
+            InvoiceRowRepository::new(&connection).upsert_one(&new_invoice)?;
+
+            let invoice_line_repo = InvoiceLineRowRepository::new(&connection);
+
+            for line in input.outbound_return_lines {
+                insert_stock_out_line(
+                    ctx,
+                    InsertStockOutLine {
+                        id: line.id.clone(),
+                        invoice_id: new_invoice.id.clone(),
+                        stock_line_id: line.stock_line_id.clone(),
+                        number_of_packs: line.number_of_packs.clone(),
+                        note: Some(line.note.clone()),
+                        r#type: Some(StockOutType::OutboundReturn),
+                        tax: None,
+                        total_before_tax: None,
+                    },
+                )
+                .map_err(|error| OutError::LineInsertError {
+                    line_id: line.id.clone(),
+                    error,
+                })?;
+
+                invoice_line_repo
+                    .update_return_reason_id(&line.id, Some(line.reason_id.clone()))
+                    .map_err(|error| OutError::LineReturnReasonUpdateError {
+                        line_id: line.id.clone(),
```

Comment:
Should there be a validation step that checks that the return reason exists? Rather than relying on the foreign key constraint error?

Reply:
Yeah i think so, especially for the correct type of reason and is_active on a reason.

This is how I would do this one:

* In generate, output (new invoice, insert_stock_out_lines, update_line_reasons), where insert_stock_out_lines = Vec<InsertStockOutLine> and update_line_reasons = Vec<UpdateReturnLineReason>
* Add a service method that will be common for inbound/outbound return line updates, and use it here, new error type will be added UpdateReturnLineReasonError, and used similar to InsertStockOutLine.

Please let me know if that makes sense or not, I think it's a good idea to do in this PR